### ISSUE TITLE
Simplify LocalDate(Time) constructor parameter names

### ISSUE THREAD

**Issue by fluidsonic** [Kotlin/kotlinx-datetime#84]
The parameter names are very weird for day-to-day use:

```kt
LocalDateTime(year = 2021, monthNumber = 1, dayOfMonth = 1, hour = 12)
```

I don't care that `month` is a number. That's what we have a type system for.
`OfMonth` for `day` is also redundant as that's obvious from the context.

https://github.com/Kotlin/kotlinx-datetime/blob/3af71d2e874592fc70282de441a09d024dcefb54/core/common/src/LocalDateTime.kt#L50

**Comment by ilya-g** [comment]
These parameters are named this way, so that one can get back the values passed to the constructed from the same-named properties of the constructed object.

**Comment by fluidsonic** [comment]
That property `dayOfMonth` should probably also simply be called `day`.

Kotlin tries to be pragmatic for the majority of use cases. The vast majority when working with dates is using the day of month. Everybody and their mother knows what a day refers to when dealing with a year/month/day date (+ time). `dayOfWeek` and `dayOfYear` is derived information and the latter used very rarely in my observation. `dayOfMonth` is overly specific. With auto-completion you also clearly see that there's `day`, `dayOfWeek` and `dayOfYear` so there's little room for error.

So, if Kotlin is a language optimized for general application development and the majority of use cases, `day` should be totally sufficient in both cases, constructor and property.

**Comment by fluidsonic** [comment]
Btw, the compiler makes things more confusing. I've provided year, month and day all as `Int` and with **natural names**.
The compiler complains that I have to provide three `Int`s. But I did.

![image](https://user-images.githubusercontent.com/85242/102360065-24f40f80-3fe4-11eb-90c9-2b29e0f112ad.png)

I also think that the names being more natural overweighs the benefit of having `monthNumber` symmetric with the property. Nobody says "month number" in day-to-day life, not even a developer. As for `dayOfMonth` see my previous reply.

**Comment by dkhalanskyjb** [comment]
What is an abomination is using the `day` field to store the index of the week in the month (https://docs.microsoft.com/en-us/windows/win32/api/timezoneapi/ns-timezoneapi-time_zone_information, see the `StandardDate` section). I agree we shouldn't provide any support for *that* kind of abominations. Zero-based months are fine by comparison and quite ubiquitous: https://grep.app/search?q=JANUARY%20%3D%200 https://grep.app/search?q=%22JANUARY%22%3A%200&words=true APIs are messy and often done as an afterthought (producing such abominations), so data plumbing benefits from getting rid of raw numbers as soon as possible in favor of typed arguments.

**Comment by dkhalanskyjb** [comment]
Here's the proposal:

* Rename `dayOfMonth` to `day` everywhere. It's consistent with how we have `LocalTime.nanosecond` and don't call it `nanosecondOfSecond` (as opposed to `nanosecondOfDay`, for example).
* Remove `monthNumber` from `LocalDate`: it's easy to access it as `month.number`, which also guides people to just use `month`. There's some code out there like `Month(date.monthNumber)`, for example.
* Rename the `monthNumber` constructor parameter to just `month`.
* Keep `monthNumber` in `DateTimeComponents` (as it's significantly different from `month` there and has its own uses).
* Keep `monthNumber` in the format builders, as having `month` + `monthName` is less clear than `monthNumber` + `monthName`.

**Comment by chrisjenx** [comment]
This seems like a change introducing ambiguity, kotlin having param names is a super nice feature, changing to  `day` instead of `dayOfMonth` feels like a step back, would have been more positive if y'all decided to use inline classes to type these, but `day` can mean anything when looking at a PR

**Comment by dkhalanskyjb** [comment]
> Peeps this defensive attitude wastes both of our time, perhaps it would be more productive to try and understand the concerns being raised to you instead of being passive-aggressive?

Of course, we are always trying to understand concerns. Your concern, according to your original message, was about `dayOfMonth` parameters (passing the day-of-month to functions):

> kotlin having param names is a super nice feature

and

> if y'all decided to use inline classes to type these

So, my answer was a direct, genuine reply to your message, and I stand by it: I still don't see how the named parameter names suffered from this change.

> Formatting sub sections of dates, this one is now not clear

Noted, thanks.

**Comment by chrisjenx** [comment]
Another example:
```
startLocalDateTime?.dayOfMonth
```

Is now:

```
startLocalDateTime?.day
```

Comes back to; is that `dayOfMonth`, `dayOfWeek`, `dayOfYear`?
